### PR TITLE
feat: seperate target address and the code address

### PR DESCRIPTION
**Pull request by bajpai244**
## Pull Request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The separation of target address and the code address are not supported, this PR adds support for this and fixed the bug to resolve `to` address correctly.

Resolves: #551 #552 

## What is the new behavior?

- target and code address has been seperated
- the correct `to` address is now resolved.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/evm/src/call_helpers.cairo [hunk 1]
```diff
@@ -60,7 +73,7 @@ impl MachineCallHelpersImpl of MachineCallHelpers {
         let caller = match call_type {
             CallType::Call => self.address(),
             CallType::DelegateCall => self.call_ctx().caller,
-            CallType::CallCode => self.address(),
+            CallType::CallCode => self.call_ctx().caller(),
```
Comment (Eikix) [inline crates/evm/src/call_helpers.cairo:76]: Are you sure about this?
https://github.com/ethereum/execution-specs/blob/3fe6514f2d9d234e760d11af883a47c1263eff51/src/ethereum/shanghai/vm/instructions/system.py#L468
https://github.com/ethereum/execution-specs/blob/3fe6514f2d9d234e760d11af883a47c1263eff51/src/ethereum/shanghai/vm/instructions/system.py#L266

Comment (bajpai244) [reply crates/evm/src/call_helpers.cairo:76]: This is a mistake, nice catch!

Comment (bajpai244) [reply crates/evm/src/call_helpers.cairo]: fixed via [this](https://github.com/kkrt-labs/kakarot-ssj/pull/553/commits/8729d01639954bf30502394c7ed4973e4c184551) commit.

**Review by Eikix — APPROVED** [review]
lgtm

**Review by enitrat — APPROVED** [review]
:lgtm:

**Pull request merged**